### PR TITLE
[WIP] 'tune cat' command for pretty printing configuration files

### PR DESCRIPTION
**Pull request by Ankur-singh**
#### Context
What is the purpose of this PR? Is it to
- [X] add a new feature
- [ ] fix a bug
- [ ] update tests and/or documentation
- [ ] other (please add here)

Please link to any issues this PR addresses. #2281 

#### Changelog
What are the changes made in this PR?
* Implemented `cat` subcommand for the CLI
* Wrote basic test-cases for `tune cat` 

#### Test plan
Please make sure to do each of the following if applicable to your PR. If you're unsure about any one of these just ask and we will happily help. We also have a [contributing page](https://github.com/pytorch/torchtune/blob/main/CONTRIBUTING.md) for some guidance on contributing.

- [X] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [X] add [unit tests](https://github.com/pytorch/torchtune/tree/main/tests/torchtune) for any new functionality
- [ ] update [docstrings](https://github.com/pytorch/torchtune/tree/main/docs/source) for any new or updated methods or classes
- [X] run unit tests via `pytest tests`
- [ ] run recipe tests via `pytest tests -m integration_test`
- [X] manually run any new or modified recipes with sufficient proof of correctness
- [X] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

When file doesn't exist:
```bash
(tune) ankur@nuc:~/github/torchtune$ tune cat llama4/8B_full.yaml
usage: tune cat [-h] config_name
tune cat: error: Config 'llama4/8B_full.yaml' not found.
```

When some other file is passed:
```bash
(tune) ankur@nuc:~/github/torchtune$ tune cat llama4/8B_full.json
usage: tune cat [-h] config_name
tune cat: error: Invalid config format: 'llama4/8B_full.json'. Must be YAML (.yaml/.yml)
```

When recipe name is passed:
```bash
(tune) ankur@nuc:~/github/torchtune$ tune cat quantize
'quantize' is a recipe, not a config. Please use a config name.
```

When correct config is passed:
```bash
(tune) ankur@nuc:~/github/torchtune$ tune cat llama2/7B_full
output_dir: /tmp/torchtune/llama2_7B/full
tokenizer:
    _component_: torchtune.models.llama2.llama2_tokenizer
    path: /tmp/Llama-2-7b-hf/tokenizer.model
    max_seq_len: null
dataset:
    _component_: torchtune.datasets.alpaca_dataset
    packed: false
seed: null
shuffle: true
model:
    _component_: torchtune.models.llama2.llama2_7b
checkpointer:
    _component_: torchtune.training.FullModelHFCheckpointer
    checkpoint_dir: /tmp/Llama-2-7b-hf
    checkpoint_files:
    - pytorch_model-00001-of-00002.bin
    - pytorch_model-00002-of-00002.bin
    recipe_checkpoint: null
    output_dir: ${output_dir}
    model_type: LLAMA2
resume_from_checkpoint: false
batch_size: 2
epochs: 1
optimizer:
    _component_: torch.optim.AdamW
    fused: true
    lr: 2e-5
loss:
    _component_: torchtune.modules.loss.CEWithChunkedOutputLoss
max_steps_per_epoch: null
gradient_accumulation_steps: 1
compile: false
optimizer_in_bwd: false
device: cuda
enable_activation_checkpointing: true
enable_activation_offloading: false
dtype: bf16
metric_logger:
    _component_: torchtune.training.metric_logging.DiskLogger
    log_dir: ${output_dir}/logs
log_every_n_steps: 1
log_peak_memory_stats: true
profiler:
    _component_: torchtune.training.setup_torch_profiler
    enabled: false
    output_dir: ${output_dir}/profiling_outputs
    cpu: true
    cuda: true
    profile_memory: false
    with_stack: false
    record_shapes: true
    with_flops: false
    wait_steps: 5
    warmup_steps: 3
    active_steps: 2
    num_cycles: 1
```

#### UX
If your function changed a public API, please add a dummy example of what the user experience will look like when calling it.
Here is a [docstring example](https://github.com/pytorch/torchtune/blob/6a7951f1cdd0b56a9746ef5935106989415f50e3/torchtune/modules/vision_transformer.py#L285)
and a [tutorial example](https://pytorch.org/torchtune/main/tutorials/qat_finetune.html#applying-qat-to-llama3-models)

- [ ] I did not change any public API
- [X] I have added an example to docs or docstrings

### CODE REVIEW

**Comment by Ankur-singh** [comment]
Still a work in progress, but I’d love to gather as much feedback as possible while it's still in the oven!

cc @RdoubleA @joecummings

**Comment by SalmanMohammadi** [comment]
Thanks so much for the awesome RFC and for putting this up @Ankur-singh!

Would you be able to attach some example output(s) from using the command to the PR description?

**Comment by Ankur-singh** [comment]
@SalmanMohammadi Thank you for the comment. I have added the command outs for different scenarios. This should make it easy for other as well. Again, thank you very much.

File: torchtune/_cli/cat.py [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+                    $ tune cat llama2/7B_full
+                    model: llama2
+                    size: 7B
+                    task: full_finetune
+                    ...
+
+                    $ tune cat non_existent_config
+                    Config 'non_existent_config' not found.
+
+                    $ tune cat some_recipe
```
Comment (joecummings) [inline torchtune/_cli/cat.py:43]: I don't think we need an example for recipe

File: torchtune/_cli/cat.py [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+                    'some_recipe' is a recipe, not a config. Please use a config name.
```
Comment (joecummings) [inline torchtune/_cli/cat.py:44]: Can you add an additional line down here saying that they can find all the `cat`-able configs via the `tune ls` command? You can also provide an example of launching a run overriding a key that was found via the new `tune cat` command.

File: torchtune/_cli/cat.py [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+
+from pathlib import Path
+from typing import List, Optional
+
+import torchtune
+import yaml
+from torchtune._cli.subcommand import Subcommand
+from torchtune._recipe_registry import Config, get_all_recipes
+
+ROOT = Path(torchtune.__file__).parent.parent
```
Comment (joecummings) [inline torchtune/_cli/cat.py:18]: Ugh, we need to find a different way to do this (laziness on my part, I apologize). I suspect the call to `import torchtune` significantly slows down the execution of the CLI. 

The key part here is that we want to find the underlying project in which the torchtune package is installed. For a local installation AND for a PyPI installation, it *should* be two levels up from the __init__.py file. So theoretically it's three levels up from any of the CLI files.

File: torchtune/_cli/cat.py [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+        return [recipe.name for recipe in get_all_recipes()]
+
+    def _get_config(self, config_str: str) -> Optional[Config]:
+        # Search through all recipes
+        for recipe in get_all_recipes():
+            for config in recipe.configs:
+                if config.name == config_str:
+                    return config
+
+    def _print_file(self, file: str) -> None:
```
Comment (joecummings) [inline torchtune/_cli/cat.py:64]: `_print_yaml_file` is a little more descriptive.

**Review by joecummings — COMMENTED** [review]
Command looks good, just a few comments on docstrings and impl

File: torchtune/_cli/cat.py [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+class Cat(Subcommand):
+    """Holds all the logic for the `tune cat` subcommand."""
+
+    def __init__(self, subparsers: argparse._SubParsersAction):
+        super().__init__()
+        self._parser = subparsers.add_parser(
+            "cat",
+            prog="tune cat",
+            help="Pretty print a config",
+            description="Pretty print a config",
```
Comment (joecummings) [inline torchtune/_cli/cat.py:30]: Can you add something related to WHY someone might want to do this? E.g. pretty print a config, making it easy to know which keys you can override when launching a training run (or something a little more concise)

File: torchtune/_cli/cat.py [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+            epilog=textwrap.dedent(
+                """\
+                examples:
+                    $ tune cat llama2/7B_full
+                    model: llama2
+                    size: 7B
+                    task: full_finetune
+                    ...
+
+                    $ tune cat non_existent_config
```
Comment (joecummings) [inline torchtune/_cli/cat.py:40]: No need for negative examples

File: torchtune/_cli/cat.py [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+        super().__init__()
+        self._parser = subparsers.add_parser(
+            "cat",
+            prog="tune cat",
+            help="Pretty print a config",
+            description="Pretty print a config",
+            epilog=textwrap.dedent(
+                """\
+                examples:
+                    $ tune cat llama2/7B_full
```
Comment (joecummings) [inline torchtune/_cli/cat.py:34]: Can you add an example of `tune cat LOCALFILE.yaml`

File: torchtune/_cli/cat.py [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+    def _print_file(self, file: str) -> None:
+        try:
+            with open(file, "r") as f:
+                data = yaml.safe_load(f)
+                if data:
+                    print(
+                        yaml.dump(
+                            data,
+                            default_flow_style=False,
+                            sort_keys=False,
```
Comment (SalmanMohammadi) [inline torchtune/_cli/cat.py:73]: I personally find the sorted configs easier to read when printed.

**Comment by SalmanMohammadi** [comment]
Would you be up for adding a little section for your shiny new command in the docs? : )

https://github.com/pytorch/torchtune/blob/main/docs/source/tune_cli.rst

File: torchtune/_cli/cat.py (hunk 4 above)
Comment (Ankur-singh) [reply torchtune/_cli/cat.py:64]: Can you please it further? What do you mean by "descriptive"?

File: torchtune/_cli/cat.py (hunk 2 above)
Comment (Ankur-singh) [reply torchtune/_cli/cat.py:44]: Thats a great idea. In fact, this is was the original motivation.

File: torchtune/_cli/cat.py (hunk 8 above)
Comment (SalmanMohammadi) [reply torchtune/_cli/cat.py:73]: cc @ebsmothers @joecummings thoughts?

File: torchtune/_cli/cat.py (hunk 4 above)
Comment (joecummings) [reply torchtune/_cli/cat.py:64]: It's a nit for sure, but this doesn't just print *any* file. It only takes in and prints a YAML file, therefore the name should be something like `print_yaml_file`.

File: torchtune/_cli/cat.py (hunk 8 above)
Comment (joecummings) [reply torchtune/_cli/cat.py:73]: I like the idea of a `--sort` argument. I agree that sorted configs *are* a little easier to read; however, they wouldn't look like our configs on Github, which would be confusing so the default should be to match what we show on Github.

File: torchtune/_cli/cat.py (hunk 3 above)
Comment (Ankur-singh) [reply torchtune/_cli/cat.py:18]: Yes, importing `torchtune` is slow. I will address this issue. We will need to update other CLI files that import `torchtune` as well, but it might be better to handle that in a separate PR to avoid scope creep.

File: torchtune/_cli/cat.py (hunk 4 above)
Comment (Ankur-singh) [reply torchtune/_cli/cat.py:64]: Yes, thats a good point. Will update the function name.

**Comment by Ankur-singh** [comment]
@SalmanMohammadi and @joecummings Thank you so much for all the amazing feedback. I have made all the requested changes.

File: torchtune/_cli/cat.py (hunk 3 above)
Comment (joecummings) [reply torchtune/_cli/cat.py]: Yep, let's do this in another PR.

File: torchtune/_cli/cat.py [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+                        max_seq_len: null
+                    ...
+
+                    # Pretty print the config in sorted order
+                    $ tune cat llama2/7B_full --sort
+
+                    # Pretty print the contents of LOCALFILE.yaml
+                    $ tune cat LOCALFILE.yaml
+
+                    # Example of launching a run overriding a key found via the `tune cat` command:
```
Comment (joecummings) [inline torchtune/_cli/cat.py:47]: super nit, but take a look at how I did this for `tune ls`: https://github.com/pytorch/torchtune/blob/d4465c8b21629949a404f821528badb35962502f/torchtune/_cli/ls.py#L37

It should be unindented OUT of the examples section and instead be considered a followup. 

e.g.

```
examples:
	$ tune cat llama2/7B_full
	...

	$ tune cat LOCALFILE.yaml
	...

You can now easily override a key based on your findings from `tune cat`:
	$ tune run full_finetune_distributed ...

Need to find all the "cat"-able configs? Try `tune ls`!
```

File: torchtune/_cli/cat.py [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+
+class Cat(Subcommand):
+    """Holds all the logic for the `tune cat` subcommand."""
+
+    def __init__(self, subparsers: argparse._SubParsersAction):
+        super().__init__()
+        self._parser = subparsers.add_parser(
+            "cat",
+            prog="tune cat",
+            help="Pretty print a config, making it easy to override parameters when using `tune run`",
```
Comment (joecummings) [inline torchtune/_cli/cat.py:28]: nit: "making it easy to *know* which parameters you can override with `tune run`.

**Review by joecummings — COMMENTED** [review]
This looks awesome! Just two more nits I swear :)

File: torchtune/_cli/cat.py (hunk 9 above)
Comment (Ankur-singh) [reply torchtune/_cli/cat.py]: Done!

File: torchtune/_cli/cat.py (hunk 10 above)
Comment (Ankur-singh) [reply torchtune/_cli/cat.py]: Done

**Review by joecummings — APPROVED** [review]
🫡🫡🫡
